src/cmds/qsub.c

```diff
@@ -227,8 +227,12 @@ static char server_out[PBS_MAXSERVERNAME + PBS_MAXPORTNUM + 2]; /* Destination s
 static struct batch_status *ss = NULL;
 static char *dfltqsubargs = NULL; /* Default qsub arguments */
 static int sd_svr; /* return from pbs_connect */
-static char script_tmp[MAXPATHLEN + 1] = ""; /* name of script file copy */
-static char fl[2 * MAXPATHLEN + 1]; /* the filename used as the pipe name */
+static char script_tmp[MAXPATHLEN + 1] = {'\0'}; /* name of script file copy */
+#ifdef WIN32
+static char fl[(2 * MAXPATHLEN) + 1] = {'\0'}; /* the filename used as the pipe name */
+#else
+static char fl[sizeof(((struct sockaddr_un *)0)->sun_path)] = {'\0'}; /* the filename used as the pipe name */
```

Comment:
This is how you get the size of a member of a structure without declaring it. In the do_daemon_stuff() function we declare a sockaddr_un structure and write to s_un.sun_path. It had to be done this way to ensure the buffer was sized appropriately.

Reply:
Why would we want to allocate more space than the sockaddr_un structure allows? What if the size of sun_path changed someday? Here's the compiler warning this change addresses:

../../repo/src/cmds/qsub.c: In function ‘daemon_submit’:
../../repo/src/cmds/qsub.c:6047:51: warning: ‘%s’ directive output may be truncated writing up to 2048 bytes into a region of size 108 [-Wformat-truncation=]
   snprintf(s_un.sun_path, sizeof(s_un.sun_path), "%s", fl);
                                                   ^~   ~~
../../repo/src/cmds/qsub.c:6047:3: note: ‘snprintf’ output between 1 and 2049 bytes into a destination of size 108
   snprintf(s_un.sun_path, sizeof(s_un.sun_path), "%s", fl);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../repo/src/cmds/qsub.c: In function ‘do_daemon_stuff’:
../../repo/src/cmds/qsub.c:5778:50: warning: ‘%s’ directive output may be truncated writing up to 2048 bytes into a region of size 108 [-Wformat-truncation=]
  snprintf(s_un.sun_path, sizeof(s_un.sun_path), "%s", fl);
                                                  ^~   ~~
../../repo/src/cmds/qsub.c:5778:2: note: ‘snprintf’ output between 1 and 2049 bytes into a destination of size 108
  snprintf(s_un.sun_path, sizeof(s_un.sun_path), "%s", fl);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reply:
I'm confused how this will cause a warning.  I suggested we change sizeof(((struct sockaddr_un *)0)->sun_path to sizeof(struct sockaddr_un).  If the compiler doesn't complain copying something into a 108 byte buffer, then it won't complain writing into a 110 byte buffer.

Reply:
Here's the warning the compiler emits when your suggestion is applied...

../../repo/src/cmds/qsub.c: In function ‘daemon_submit’:
../../repo/src/cmds/qsub.c:6052:51: warning: ‘%s’ directive output may be truncated writing up to 109 bytes into a region of size 108 [-Wformat-truncation=]
   snprintf(s_un.sun_path, sizeof(s_un.sun_path), "%s", fl);
                                                   ^~   ~~
../../repo/src/cmds/qsub.c:6052:3: note: ‘snprintf’ output between 1 and 110 bytes into a destination of size 108
   snprintf(s_un.sun_path, sizeof(s_un.sun_path), "%s", fl);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../repo/src/cmds/qsub.c: In function ‘do_daemon_stuff’:
../../repo/src/cmds/qsub.c:5783:50: warning: ‘%s’ directive output may be truncated writing up to 109 bytes into a region of size 108 [-Wformat-truncation=]
  snprintf(s_un.sun_path, sizeof(s_un.sun_path), "%s", fl);
                                                  ^~   ~~
../../repo/src/cmds/qsub.c:5783:2: note: ‘snprintf’ output between 1 and 110 bytes into a destination of size 108
  snprintf(s_un.sun_path, sizeof(s_un.sun_path), "%s", fl);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reply:
Oh, that is true.  If we make the buffer 110 bytes, it will complain about copying it into a 108 byte buffer.  Unfortunately we'll have to deal with the somewhat confusing original form.